R/dataframe-frame.R

```diff
@@ -112,16 +113,55 @@ dataframe__to_struct <- function(name = "") {
     wrap()
 }
 
+#' Convert an existing DataFrame to a LazyFrame
+#' @description Start a new lazy query from a DataFrame.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' pl$DataFrame(iris)$lazy()
 dataframe__lazy <- function() {
   self$`_df`$lazy() |>
     wrap()
 }
 
+#' Clone a DataFrame
+#'
+#' This makes a very cheap deep copy/clone of an existing
+#' [`DataFrame`]. Rarely useful as `DataFrame`s are nearly 100%
+#' immutable. Any modification of a `DataFrame` should lead to a clone anyways,
+#' but this can be useful when dealing with attributes (see examples).
+#'
+#' @inherit as_polars_df return
+#' @examples
+#' df1 <- pl$DataFrame(iris)
+#'
+#' # Make a function to take a DataFrame, add an attribute, and return a DataFrame
+#' give_attr <- function(data) {
+#'   attr(data, "created_on") <- "2024-01-29"
+#'   data
+#' }
+#' df2 <- give_attr(df1)
+#'
+#' # Problem: the original DataFrame also gets the attribute while it shouldn't!
+#' attributes(df1)
+#'
+#' # Use $clone() inside the function to avoid that
+#' give_attr <- function(data) {
+#'   data <- data$clone()
+#'   attr(data, "created_on") <- "2024-01-29"
+#'   data
+#' }
+#' df1 <- pl$DataFrame(iris)
+#' df2 <- give_attr(df1)
+#'
+#' # now, the original DataFrame doesn't get this attribute
+#' attributes(df1)
```

Comment:
I think these examples are not good...

Reply:
This is actually the only case that I could find where `clone()` is useful (the python docs don't explain at all why one would need to use it).

Reply:
The behavior of renaming a column using `.columns = ` in Python Polars is a good example, but that is not implemented here (instead we have this)...

https://github.com/eitsupi/neo-r-polars/blob/557325983d62686c0d35dc820e0ff2af1310e916/R/dataframe-frame.R#L88-L95

Unlike the current R polars, here the DataFrame is an environment object, so how about an example like this? (i.e., environments are one of the few objects in R that are passed by reference)

```r
df_a <- pl$DataFrame(x = 1)
df_b <- df_a
df_c <- df_a$clone()
rlang::env_label(df_a)
rlang::env_label(df_b)
rlang::env_label(df_c)
```

Reply:
Yes but that doesn't show why one would need this. I don't mind putting this `rlang::env_label()` before the other example to show that the address changes, but I'd also like to show a valid usecase for this and I think the one with attributes does the job. What don't you like about it? I don't mind replacing them if you have some alternative usecases

Reply:
What is the reason for adding the attribute?
I don't understand the use case for adding attributes without defining another class.
If you add an attribute, it will disappear immediately upon some operation.

Reply:
Here's how I use it in `group_by`: https://github.com/etiennebacher/tidypolars/blob/main/R/groups.R#L31

Reply:
Anyway, since this is the behavior of environment types in general, not just polars object, can you add examples of `rlang::env_label` first?
Then as a developmental example you can explain why it matters.

Reply:
> Will using attributes affect print results, etc.?

I have a custom print method for `tidypolars` objects so I handle attributes there.

I'll add the rlang::env_label() examples